File: src/main/java/com/indiepost/controller/admin/AdminLinkController.java
package com.indiepost.controller.admin;

import com.indiepost.dto.stat.LinkDto;
import com.indiepost.service.LinkService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import javax.validation.ValidationException;
import java.util.List;
import java.util.Objects;

/**
 * Created by jake on 8/10/17.
 */
@RestController
@RequestMapping(value = "/admin/link", produces = {"application/json; charset=UTF-8"})
public class AdminLinkController {

    private final LinkService linkService;

    @Autowired
    public AdminLinkController(LinkService linkService) {
        this.linkService = linkService;
    }

    @GetMapping
    public List<LinkDto> getLinks() {
        return linkService.findAll();
    }

    @GetMapping(value = "/{id}")
    public LinkDto getLink(@PathVariable Long id) {
        return linkService.findById(id);
    }

    @PostMapping
    public LinkDto createLink(@Valid @RequestBody LinkDto linkDto) {
        return linkService.save(linkDto);
    }

    @PutMapping(value = "/{id}")
    public void updateLink(@PathVariable Long id, @Valid @RequestBody LinkDto linkDto) {
        if (!Objects.equals(id, linkDto.getId())) {
            throw new ValidationException("REST resource ID and LinkDto::id are not match.");
        }
        linkService.update(linkDto);
    }

    @DeleteMapping(value = "/{id}")
    public void deleteLink(@PathVariable Long id) {
        linkService.deleteById(id);
    }
}

File: src/main/kotlin/com/indiepost/controller/admin/AdminLinkController.kt
package com.indiepost.controller.admin

import com.indiepost.dto.LinkMetadataResponse
import com.indiepost.dto.analytics.LinkDto
import com.indiepost.service.LinkService
import org.springframework.web.bind.annotation.*
import javax.inject.Inject
import javax.validation.Valid
import javax.validation.ValidationException

/**
 * Created by jake on 8/10/17.
 */
@RestController
@RequestMapping(value = ["/admin/link"], produces = ["application/json; charset=UTF-8"])
class AdminLinkController @Inject
constructor(private val linkService: LinkService) {

    @GetMapping("/{id}")
    fun getLink(@PathVariable id: Long): LinkDto? {
        return linkService.findById(id)
    }

    @PostMapping
    fun createLink(@Valid @RequestBody linkDto: LinkDto): LinkDto {
        return linkService.save(linkDto)
    }

    @GetMapping("/search/film")
    fun searchFilms(@RequestParam text: String, @RequestParam size: Int): List<LinkMetadataResponse> {
        return linkService.searchMovies(text, size)
    }

    @GetMapping("/search/book")
    fun searchBooks(@RequestParam text: String, @RequestParam size: Int): List<LinkMetadataResponse> {
        return linkService.searchBooks(text, size)
    }

    @GetMapping("/from-url")
    fun fromUrl(@RequestParam url: String): LinkMetadataResponse? {
        return linkService.getFromUrl(url)
    }

    @PutMapping("/{id}")
    fun updateLink(@PathVariable id: Long, @Valid @RequestBody linkDto: LinkDto) {
        if (id != linkDto.id) {
            throw ValidationException("REST resource ID and LinkDto::id are not match.")
        }
        linkService.update(linkDto)
    }

    @DeleteMapping("/{id}")
    fun deleteLink(@PathVariable id: Long) {
        linkService.deleteById(id)
    }
}
